plugins/flytekit-docker/flytekitplugins/docker/image_builder.py

```diff
@@ -0,0 +1,81 @@
+import os
+import subprocess
+import tempfile
+
+import click
+
+from flytekit.image_spec.image_spec import _F_IMG_ID, ImageBuildEngine, ImageSpec, ImageSpecBuilder
+
+
+class DockerfileImageSpecBuilder(ImageSpecBuilder):
+    """
+    This class is used to build a docker image using docker
+    """
+
+    def execute_command(self, command: list[str]) -> None:
+        click.secho(f"Run command: {' '.join(command)} ", fg="blue")
+        p = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        for line in iter(p.stdout.readline, ""):
+            if p.poll() is not None:
+                break
+            line_out = line.decode().strip()
+            if line.decode().strip() != "":
+                click.secho(line_out, fg="blue")
+
+        if p.returncode != 0:
+            _, stderr = p.communicate()
+            raise Exception(f"failed to run command {command} with error {stderr}")
+
+    def build_image(self, image_spec: ImageSpec):
+        if image_spec.dockerfile is None:
+            raise RuntimeError("Image spec dockerfile cannot be None")
+        if image_spec.platform is None:
+            raise RuntimeError("Image spec platform cannot be None")
+        if image_spec.source_root is None:
+            raise RuntimeError("Image spec source_root cannot be None")
+        env_image_id = os.environ.get(_F_IMG_ID)
+        if env_image_id:
+            click.secho(f"Skipping nested build of {image_spec.name}", fg="blue")
+
+            def horrible_horrible_hack() -> str:
+                return env_image_id
+
+            image_spec.image_name = horrible_horrible_hack
+            return
+
+        # Inject _F_IMG_ID to stop recursive builds
```

Comment:
You say:

> horrible_horrible_hack is because the Dockerfile isn't packaged with the script and sent to the remote. So when the remote calls .build_image and tries to determine a hash, it either crashes (folder doesn't exist) or it determines a new one and thus returns a new image_name. This causes the builder to try and build the image again and then crash because docker doesn't exist. 

With `when the remote calls .build_image and tries to determine a hash`, do you mean that this happens in the task pod or during the registration process on the user's machine? Sorry for my confusion.